test.cpp

```diff
@@ -238,6 +238,21 @@ TEST(HostnameTest, LowercasesHostname)
     EXPECT_EQ("www.testing.com", Url::Url("http://www.testing.com/FOO").host());
 }
 
+TEST(QueryTest, SanitizesQuery)
+{
+    EXPECT_EQ("a=1&b=2"    , Url::Url("http://foo.com/?a=1&&&&&&b=2"   ).query());
+    EXPECT_EQ("foo=2"      , Url::Url("http://foo.com/????foo=2"       ).query());
+    EXPECT_EQ("foo=2"      , Url::Url("http://foo.com/?foo=2&&&"       ).query());
+}
```

Comment:
It looks like your code won't handle this case: `""http://foo.com/?a=1&?&?&?b=2"` because the duplicates aren't adjacent.